File: python/802_找到最终的安全状态_medium.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        color = [0] * n
        def safe(x):
            if color[x] > 0:
                return color[x] == 2
            color[x] = 1
            for y in graph[x]:
                if not safe(y):
                    return False
            color[x] = 2
            return True
        return [i for i in range(n) if safe(i)]

File: cpp/802_找到最终的安全状态_medium.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> ans;
        vector<int> color(n, 0);

        function<bool(int)> safe = [&](int x) {
            if (color[x] > 0)
                return color[x] == 2;
            color[x] = 1;
            for (int y : graph[x]) {
                if (!safe(y))
                    return false;
            }
            color[x] = 2;
            return true;
        };

        for(int i=0; i<n; i++)
            if (safe(i))
                ans.emplace_back(i);
        return ans;
    }
};